bootstrap/kubeadm/api/v1alpha3/kubeadmbootstrapconfig_types.go

```diff
@@ -175,6 +183,71 @@ type File struct {
 	Content string `json:"content"`
 }
 
+// FileSource is a union of all possible external source types for file data.
+// Only one field may be populated in any given instance. Developers adding new
+// sources of data for target systems should add them here.
+type FileSource struct {
+	// Secret represents a secret that should populate this file.
+	// +optional
+	Secret *SecretFileSource `json:"secret,omitempty"`
+}
+
+type FileSourceOptions struct {
+	// DefaultPermissions specifies optional mode bits to use on created files by default.
+	// Must be a value between 0 and 0777.
+	// +optional
+	DefaultPermissions string `json:"defaultPermissions,omitempty"`
+
+	// DefaultOwner specifies the ownership of the file, e.g. "root:root".
+	// +optional
+	DefaultOwner string `json:"defaultOwner,omitempty"`
+
+	// DefaultEncoding specifies the encoding of the file contents.
+	// +optional
+	DefaultEncoding Encoding `json:"defaultEncoding,omitempty"`
+}
+
+// Adapts a Secret into a FileSource.
+//
+// The contents of the target Secret's Data field will be presented
+// as files using the keys in the Data field as the file names.
+type SecretFileSource struct {
+	FileSourceOptions `json:",inline"`
+
+	// Name of the secret in the KubeadmBootstrapConfig's namespace to use.
+	SecretName string `json:"secretName"`
+
+	// If unspecified, each key-value pair in the Data field of the referenced
+	// Secret will be projected as a file whose name is the
+	// key and content is the value. If specified, the listed keys will be
+	// projected into the specified paths, and unlisted keys will not be
+	// present. If a key is specified which is not present in the Secret,
+	// the volume setup will error.
+	// +optional
+	Items []KeyToPath `json:"items,omitempty"`
```

Comment:
This effectively gives a way to users to do the same thing in multiple ways. Should we instead find a single solution, at least for now, that by contract has 1 secret for each file you'd want to inject?

A few benefits:
- API would be simpler (no need for defaults, and defaults for those defaults).
- Secrets wouldn't be able to grow too much (etcd storage limitations).
- Clear separation between each secret.
- No need to look at keys, we'd force each secret to have a single `value` secret.

Reply:
quick note before the rest of my response, I discovered yesterday that secret data keys don't map well to file paths (can't use slashes like in config maps). So we definitely need some form of metadata to map the data keys into file paths.

I think 1 value per source is okay, I do think there's a case for reuse that will come up eventually. I don't have a firm scenario right now for Azure, as azure.json is 1 file per cluster, and 1 per secret currently. 

After trying this, pattern *might* emerge (at least for Azure) for users to have 1 secret per cluster with all the sensitive data fully spec'd out, and they don't have to jump around with references. This would happen kind of naturally since we'd already need 1 secret per cluster due to the uniqueness of cloud provider file.

I think future-proofing has some benefits, given the resulting API is pretty usable for the single-file, single-key case while expanding the alternate API would be less flexible (breakage around key/item mapping. I don't really like the top level expansion of defaultPermissions, defaultEncoding, etc. -- maybe we should force that to be fully qualified?

any other thoughts? @ncdc since you were on the original discussion too?